Change navigation layout on small screens in web UI

Gargron opened the pull request:
- Removes extra header used for logo/sign-up, moves logo to Home/Explore header
- Adds bottom navigation bar with hamburger menu
- Adds profile card to top of sidebar
- Changes sidebar to float over the screen when menu is open instead of being permanently shown
- Changes icon used for "About" page in sidebar from horizontal ellipsis to info
- Changes icon used for column settings from settings to tune

|Logged in|Logged in, menu open|
|--|--|
|![Screen Shot 2025-06-07 at 06 16 53](https://github.com/user-attachments/assets/707ad1e5-8c85-480a-afd9-12d5d7c01931)|![Screen Shot 2025-06-07 at 06 16 58](https://github.com/user-attachments/assets/5937d88d-0d8e-4426-a850-7bac77959438)|

![Screen Shot 2025-06-07 at 06 17 06](https://github.com/user-attachments/assets/6c1c6214-9107-4fbc-a557-69fac3cb34cf)


___

Fixes #19571

Gargron committed:
Change navigation layout on small screens in web UI

vmstan commented:
@daprice has some great comments so I'll just add to what he has said:
- Confirm notes about the tab bar being difficult to use when setup as a PWA (Safari on iPhone) although I don't typically use it like this and the existing layout has plenty of quirks here.
- Confirm notes about scrolling when the hamburger menu is open. 
- Confirm comment on needing a way to log out on mobile. White space next to account name seems to be the most logical area.
![IMG_6786](https://github.com/user-attachments/assets/48638c13-4d28-4441-99a1-fb41d98e289f)

- I think the use of the icon is great as a replacement for the word mark logo. 
- It's also customizable by the instance as much as the word mark is/was.
![IMG_6787](https://github.com/user-attachments/assets/0ef56016-e66c-4ae8-a3a9-ff031df5ff11)

- There is no way to post a new post when the screen is between the full single column view and the narrow column.
![CleanShot 2025-06-06 at 09 15 58@2x](https://github.com/user-attachments/assets/499f2dd7-e6ff-48a2-864a-e4bb8f64e746)

Gargron commented:
- Added extra padding to the bottom bar and the menu based on safe area insets
- Added expand/collapse button to the lists panel within the menu
- Added log out button to the profile card
- Added new post button to the menu
- Changed home/explore header icons to be the logo up until the point the full logo becomes visible
- Changed Explore to Search in the menu up to the point the search input becomes visible
- Changed which element scrolls within the menu
- Changed breakpoint of when the menu becomes hidden so it remains visible on the iPad
- Fixed menu shadow showing on the side when the menu is hidden

daprice commented:
Updated my above comment to check off the issues that are fixed for me. The main thing is that scrolling in the hamburger menu doesn’t seem any better than before. The safe area inset padding works great for me in Safari but not as a PWA (I added a note about this in my other comment). Everything else looks good to me.

#34937 and #34938 also still happen on iPad, but those aren’t specific to this PR

diondiondion commented on app/javascript/mastodon/features/ui/components/navigation_bar.tsx:
```diff
@@ -0,0 +1,217 @@
+      to={to}
+      aria-label={title}
+    >
+      {match?.isExact && activeIcon ? activeIcon : icon}
+      {iconComponent && (
+        <Icon
+          id=''
+          icon={
+            match?.isExact && activeIconComponent
+              ? activeIconComponent
+              : iconComponent
+          }
+        />
+      )}
+    </NavLink>
```
I found this bit of code hard to wrap my head around as just going by the internal logic it's entirely possible for two icons to be rendered in one link which seems unexpected. Might be simpler to forego the `iconComponent`/`activeIconComponent` shortcut and always passing in `ReactNodes`

diondiondion commented on app/javascript/mastodon/features/ui/hooks/useBreakpoint.tsx:
```diff
@@ -0,0 +1,52 @@
+import { useState, useEffect } from 'react';
+
+const breakpoints = {
+  openable: 759,
```
Should this be `mobileMenuOpenable` for clarity when used in other contexts?

diondiondion commented on app/javascript/mastodon/features/ui/hooks/useBreakpoint.tsx:
```diff
@@ -0,0 +1,52 @@
+  const displayName = `withMobileLayout(${Component.displayName ?? Component.name})`;
+
+  const ComponentWithBreakpoint = (props: P) => {
+    const matchesBreakpoint = useBreakpoint('full');
```
It's not obvious that `withBreakpoint` only supports a single fixed breakpoint, but it's probably ok for now and could be extended later if necessary.

Gargron replied:
Added a way to pass the value.

diondiondion commented on app/javascript/mastodon/components/column_header.tsx:
```diff
@@ -238,7 +239,10 @@ export const ColumnHeader: React.FC<Props> = ({
-          <Icon id='sliders' icon={SettingsIcon} />
+          <Icon
+            id='sliders'
+            icon={collapsed ? UnfoldMoreIcon : UnfoldLessIcon}
```
I agree with the feedback someone else posted earlier that these icons imply an interaction where you'd select from a set of options, e.g. from Home to Explore.

diondiondion replied:
FWIW, the Twitter-like theme used on indieweb.social (forgot its name) uses a 90-degrees turned tune icon.

<img width="516" alt="image" src="https://github.com/user-attachments/assets/1fd12029-08d8-4aa6-b8f7-b1bf733fbc4b" />

diondiondion reviewed (dismissed):
LGTM, however there still seem to be a few issues around scrollability of content behind the menu, but they could be addressed in a follow-up. (I don't know if a simple `overflow: hidden` on the body can solve this on Safari iOS these days.)

Gargron commented on app/javascript/mastodon/features/ui/components/list_panel.tsx:
```diff
@@ -0,0 +1,89 @@
+          <IconButton
+            icon='down'
+            expanded={expanded}
+            iconComponent={expanded ? ArrowDropDownIcon : ArrowLeftIcon}
```
See Neue Layout in Figma for reference--I think it's this way because it's meant to mimic a tree.

diondiondion replied:
I couldn't find the "collapsed" state in Figma. I think it's just the left-facing chevron that I find odd, because the UI pattern otherwise matches the common "accordion"-style disclosure:

(Random image example:)
![image](https://github.com/user-attachments/assets/f86b2d07-fdb6-48bb-a7ca-c25ebec30c7b)

Gargron replied:
Right, I think it can be done both ways, but e.g. in Notion when you have a collapsed section the arrow is pointing towards the section title.

CyberTailor commented on app/javascript/mastodon/components/column_header.tsx:
```diff
@@ -9,7 +9,8 @@ import ArrowBackIcon from '@/material-icons/400-24px/arrow_back.svg?react';
 import ChevronLeftIcon from '@/material-icons/400-24px/chevron_left.svg?react';
 import ChevronRightIcon from '@/material-icons/400-24px/chevron_right.svg?react';
 import CloseIcon from '@/material-icons/400-24px/close.svg?react';
-import SettingsIcon from '@/material-icons/400-24px/settings.svg?react';
```
`NotificationsPermissionBanner` still uses the old icon

diondiondion replied:
Thanks for reporting, just opened a PR here: https://github.com/mastodon/mastodon/pull/35193
